2230 add language button

jacknussbaum opened the pull request:
Resolves #2230 

Adds a button to the navbar that allows users to change their language settings, refreshing the page and displaying text in their chosen language. Language preferences are then saved locally in the user's browser for later reference.

![image](https://user-images.githubusercontent.com/71039842/110683524-f9229900-8190-11eb-9df1-e2740376c3ce.png)

jacknussbaum commented:
Realized the dropdown is slightly out of place:
![image](https://user-images.githubusercontent.com/71039842/110683608-122b4a00-8191-11eb-9324-d6b61ac56124.png)

Will fix that. 

Also, Mikey, at the moment the PLAY_LANG  cookie is saved as a session cookie, meaning it is not saved for the websites use at a later date. I am looking into how to change the save duration of the cookie, but for now the page changes language and refreshes as desired!

jacknussbaum commented:
Also, feel free to give any design/placement feedback. I was more concerned with getting the functionality working at this point and didn't play around much with the visuals.

misaugstad commented:
This is great!! Glad you figured this all out!

>Realized the dropdown is slightly out of place

That gray box around the button when you click on it also doesn't look great. Could you figure out how to have that not happen? Or least be more centered?

>Also, Mikey, at the moment the PLAY_LANG cookie is saved as a session cookie, meaning it is not saved for the websites use at a later date. I am looking into how to change the save duration of the cookie

I think that what you have right now is probably what we're looking for. The best way to make it persist longer would be to save the language selection as part of a user's profile. But I think that that's out of the scope of this PR.

jonfroehlich commented:
Great work Jack!

Agree with Mikey.

I think we'll definitely want persistent language preference storage—and we
should do whatever is "best" here (I was thinking Cookie, sounds like Mikey
thinks user profile--which has benefit of working across computers).

Jon

On Wed, Mar 10, 2021 at 11:28 AM Mikey Saugstad ***@***.***>
wrote:

> This is great!! Glad you figured this all out!
>
> Realized the dropdown is slightly out of place
>
> That gray box around the button when you click on it also doesn't look
> great. Could you figure out how to have that not happen? Or least be more
> centered?
>
> Also, Mikey, at the moment the PLAY_LANG cookie is saved as a session
> cookie, meaning it is not saved for the websites use at a later date. I am
> looking into how to change the save duration of the cookie
>
> I think that what you have right now is probably what we're looking for.
> The best way to make it persist longer would be to save the language
> selection as part of a user's profile. But I think that that's out of the
> scope of this PR.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ProjectSidewalk/SidewalkWebpage/pull/2486#issuecomment-795955807>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAML55LBL2JXO55TAQRREQLTC623DANCNFSM4Y6V5IGA>
> .
>


-- 
Jon E. Froehlich <https://jonfroehlich.github.io/> (he/him
<https://www.mypronouns.org/he-him>) | @jonfroehlich
<https://twitter.com/jonfroehlich>
Associate Professor, Allen School of Computer Science & Engineering
Core Faculty, Urban Design and Planning, College of Built Environments
Director, Makeability Lab <https://makeabilitylab.cs.washington.edu/> |
Associate Director, CREATE <http://create.uw.edu/>
University of Washington
Help make sidewalks more accessible: http://projectsidewalk.io

misaugstad commented:
Sorry to do a 180 on this! But I didn't fully think this through or how easy it might be to make a persistent cookie. I think all you'd need to do is add an expiry time to the cookie... So actually feel free to go ahead and try that out! If it's easy and it works, then I'm all for it!

jacknussbaum commented:
No worries, I actually think its pretty easy typically. In my past work with js cookies, its as easy as sending one more parameter to a "setCookie" function call.

jacknussbaum commented:
Thank you both for your feedback!

Will look into changing the gray box on the button as well!

misaugstad commented:
One thing that I'm thinking about now is if we would need to provide a separate cookie consent policy to be GDPR compliant if we were to make the cookie persistent. Been reading up on the exemptions to having a separate cookie policy. Here is the easiest to understand documentation I've found so far. It's for PECR in the UK, but supposedly all country-specific cookie-related laws that came out of GDPR have very similar rules: https://ico.org.uk/for-organisations/guide-to-pecr/guidance-on-the-use-of-cookies-and-similar-technologies/how-do-we-comply-with-the-cookie-rules/#comply19

>Session cookies used to store a user's preference can rely on the strictly necessary exemption, provided they are not linked to a persistent identifier.
>
>The exemption may in some cases also apply to persistent cookies but the user must be given sufficient information in a prominent location - for example, cookies used as part of a cookie consent mechanism, which remember the user's cookie preferences over a period of time (eg 90 days), can be exempt.

^^ This language in the section on exemptions makes it sound like we might need to let them know that we are storing a persistent cookie for a user preference change..?

jacknussbaum commented:
Fixed the visuals of the button!

-Removed dashed outline on hover
-Removed gray background on button active
-Resized the dropdown

![image](https://user-images.githubusercontent.com/71039842/111011571-b4446100-834e-11eb-8497-074a4ab47d14.png)

Let me know your thoughts!

misaugstad commented:
yep, looks better!!

jonfroehlich commented:
agree. can we PR this so we can get it into the next release?

This is something that Liga Peatonal has really wanted for a while now...

On Fri, Mar 12, 2021 at 5:58 PM Mikey Saugstad ***@***.***>
wrote:

> yep, looks better!!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ProjectSidewalk/SidewalkWebpage/pull/2486#issuecomment-797846698>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAML55M4OUQ4WUDWYBEPDC3TDK2DTANCNFSM4Y6V5IGA>
> .
>


-- 
Jon E. Froehlich <https://jonfroehlich.github.io/> (he/him
<https://www.mypronouns.org/he-him>) | @jonfroehlich
<https://twitter.com/jonfroehlich>
Associate Professor, Allen School of Computer Science & Engineering
Core Faculty, Urban Design and Planning, College of Built Environments
Director, Makeability Lab <https://makeabilitylab.cs.washington.edu/> |
Associate Director, CREATE <http://create.uw.edu/>
University of Washington
Help make sidewalks more accessible: http://projectsidewalk.io

jacknussbaum commented:
Mikey, I am not quite sure what your current stance is on the cookie situation? If this is something that would be easier to store user's profile and avoid the whole user consent aspect, then this PR is definitely in a state that can be rolled out. The language change can be super easily applied on any page of the website without redirecting you. Its pretty minimally intrusive to the user experience to update the language on a session basis for now.

misaugstad commented:
Thanks jack. Yeah I’ll review as-is for now, and persistence can be addressed in the future. I’ll review later today.

Do you have availability to make edits today if I find anything? If not that’s totally fine and I’ll probably just make the edits myself so I can put this in the new release today.

jacknussbaum commented:
I will be away from my computer most of today, unfortunately. Making the edits yourself may be more expedient, thank you.

misaugstad commented:
Made some small changes which took way longer than expected :grimacing: :

1. It didn't look great in the hamburger menu when the window is narrow so I fixed that up.
2. You didn't include the `navbar-button` CSS class because we didn't want to include the border around the button, but it left out a bunch of other CSS properties we wanted to have included. I added that class to the button so that we can just override the properties that we want to change.
3. Added logging of clicks to the button.
4. Added alt text to the icon for accessibility purposes.
5. The different languages in the dropdown had the same `id` attributes, and those are supposed to be unique, so I appended the language code to the IDs.
